### PR TITLE
Add Linux Memory Metrics

### PR DESCRIPTION
**Pull request by SpriCoder**
QQ20240413-235351@2x

### CODE REVIEW

File: iotdb-core/metrics/interface/src/main/java/org/apache/iotdb/metrics/metricsets/system/SystemMetrics.java [hunk 1]
```diff
@@ -166,6 +173,48 @@ private void collectSystemMemInfo(AbstractMetricService metricService) {
         a -> osMxBean.getCommittedVirtualMemorySize(),
         SystemTag.NAME.toString(),
         SYSTEM);
+    if (CONFIG.getSystemType() == SystemType.LINUX) {
+      metricService.createAutoGauge(
```
Comment (OneSizeFitsQuorum) [inline iotdb-core/metrics/interface/src/main/java/org/apache/iotdb/metrics/metricsets/system/SystemMetrics.java:177]: add clear logic in `removeSystemMemInfo`?

File: iotdb-core/metrics/interface/src/main/java/org/apache/iotdb/metrics/metricsets/system/SystemMetrics.java [hunk 2]
```diff
@@ -166,6 +173,48 @@ private void collectSystemMemInfo(AbstractMetricService metricService) {
...
+          SystemMetric.LINUX_MEMORY_COUNT.toString(),
+          MetricLevel.CORE,
+          this,
+          a -> a.updateLinuxSystemMemInfo(metricService),
+          SystemTag.NAME.toString());
+    }
+  }
+
+  private final String[] linuxMemoryTitles =
```
Comment (OneSizeFitsQuorum) [inline iotdb-core/metrics/interface/src/main/java/org/apache/iotdb/metrics/metricsets/system/SystemMetrics.java:186]: move to the top?

File: iotdb-core/metrics/interface/src/main/java/org/apache/iotdb/metrics/metricsets/system/SystemMetrics.java [hunk 3]
```diff
@@ -166,6 +173,48 @@ private void collectSystemMemInfo(AbstractMetricService metricService) {
...
+      Process process = runtime.exec(getSystemMemoryCommand);
+      StringBuilder result = new StringBuilder();
+      try (BufferedReader input =
+          new BufferedReader(new InputStreamReader(process.getInputStream()))) {
+        String line;
+        while ((line = input.readLine()) != null) {
+          result.append(line).append("\n");
+        }
+      }
+      logger.error(result.toString());
```
Comment (OneSizeFitsQuorum) [inline iotdb-core/metrics/interface/src/main/java/org/apache/iotdb/metrics/metricsets/system/SystemMetrics.java:201]: remove this?

File: iotdb-core/metrics/interface/src/main/java/org/apache/iotdb/metrics/metricsets/system/SystemMetrics.java [hunk 4]
```diff
@@ -166,6 +173,48 @@ private void collectSystemMemInfo(AbstractMetricService metricService) {
...
+      String[] lines = result.toString().split("\n");
+      String[] memParts = lines[1].trim().split("\\s+");
```
Comment (OneSizeFitsQuorum) [inline iotdb-core/metrics/interface/src/main/java/org/apache/iotdb/metrics/metricsets/system/SystemMetrics.java:203]: if length != 7, then maybe wo do not need to collect

File: iotdb-core/metrics/interface/src/main/java/org/apache/iotdb/metrics/metricsets/system/SystemMetrics.java [hunk 5]
```diff
@@ -166,6 +173,48 @@ private void collectSystemMemInfo(AbstractMetricService metricService) {
...
+      count = memParts.length;
+      for (int i = 1; i < count; i++) {
+        metricService
+            .getOrCreateGauge(
+                SystemMetric.LINUX_MEMORY_SIZE.toString(),
+                MetricLevel.CORE,
+                SystemTag.NAME.toString(),
+                linuxMemoryTitles[i - 1])
+            .set(Long.parseLong(memParts[i]) * 1024);
```
Comment (OneSizeFitsQuorum) [inline iotdb-core/metrics/interface/src/main/java/org/apache/iotdb/metrics/metricsets/system/SystemMetrics.java:212]: do we need to multiple 1024? or we just need to set it in dashboard?

File: iotdb-core/metrics/interface/src/main/java/org/apache/iotdb/metrics/metricsets/system/SystemMetrics.java [hunk 6]
```diff
@@ -166,6 +173,48 @@ private void collectSystemMemInfo(AbstractMetricService metricService) {
...
+      }
+    } catch (IOException e) {
+      logger.warn("Failed to get memory, because ", e);
```
Comment (OneSizeFitsQuorum) [inline iotdb-core/metrics/interface/src/main/java/org/apache/iotdb/metrics/metricsets/system/SystemMetrics.java:215]: maybe debug is ok

File: iotdb-core/metrics/interface/src/main/java/org/apache/iotdb/metrics/metricsets/system/SystemMetrics.java (hunk 1 above)
Comment (OneSizeFitsQuorum) [inline iotdb-core/metrics/interface/src/main/java/org/apache/iotdb/metrics/metricsets/system/SystemMetrics.java:177]: so do we need LINUX_MEMORY_COUNT metric?

File: iotdb-core/metrics/interface/src/main/java/org/apache/iotdb/metrics/metricsets/system/SystemMetrics.java (hunk 5 above)
Comment (SpriCoder) [reply iotdb-core/metrics/interface/src/main/java/org/apache/iotdb/metrics/metricsets/system/SystemMetrics.java:212]: I think it's a good idea to keep the union of related metrics same

File: iotdb-core/metrics/interface/src/main/java/org/apache/iotdb/metrics/metricsets/system/SystemMetrics.java (hunk 1 above)
Comment (SpriCoder) [reply iotdb-core/metrics/interface/src/main/java/org/apache/iotdb/metrics/metricsets/system/SystemMetrics.java:177]: Maybe it's can be flag of how many linux memory we get

File: iotdb-core/metrics/interface/src/main/java/org/apache/iotdb/metrics/metricsets/system/SystemMetrics.java [hunk 7]
```diff
@@ -194,6 +250,15 @@ private void removeSystemMemInfo(AbstractMetricService metricService) {
         SystemMetric.SYS_COMMITTED_VM_SIZE.toString(),
         SystemTag.NAME.toString(),
         SYSTEM);
+    if (CONFIG.getSystemType() == SystemType.LINUX) {
+      for (String title : linuxMemoryTitles) {
+        metricService.remove(
+            MetricType.GAUGE,
+            SystemMetric.LINUX_MEMORY_SIZE.toString(),
```
Comment (OneSizeFitsQuorum) [inline iotdb-core/metrics/interface/src/main/java/org/apache/iotdb/metrics/metricsets/system/SystemMetrics.java:257]: clear LINUX_MEMORY_COUNT too?
Or maybe we do not need to records the count?

**Review by OneSizeFitsQuorum — APPROVED** [review]
LGTM

**Pull request merged**